firmware/kivipallur_firmware_3.0/Core/Src/main.c

```diff
@@ -0,0 +1,1231 @@
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_TIM1_Init();
+  MX_RTC_Init();
+  MX_TIM2_Init();
+  MX_TIM3_Init();
+  MX_TIM8_Init();
+  MX_TIM17_Init();
+  MX_TIM16_Init();
+  MX_TIM4_Init();
+  MX_TIM15_Init();
+  MX_USB_Device_Init();
+  MX_DMA_Init();
+  MX_ADC1_Init();
+  MX_CRC_Init();
+  MX_TIM6_Init();
+  /* USER CODE BEGIN 2 */
+
+  HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_2);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_3);
+  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_4);
+
+  HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_2);
+  HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_2);
+  HAL_TIM_Encoder_Start(&htim8, TIM_CHANNEL_1);
+  HAL_TIM_Encoder_Start(&htim8, TIM_CHANNEL_2);
+
+/*
+  TIM15->CCR2 = 3000;
+*/
+/*
+  TIM1->CCR1 = 0;
+  TIM1->CCR2 = 20000;
+  TIM2->CCR1 = 20000;
+  TIM2->CCR2 = 0;
+  TIM2->CCR3 = 0;
+  TIM2->CCR4 = 20000;
+*/
+
+  TIM1->CCR1 = 0;
+  TIM1->CCR2 = 12000;
+  TIM2->CCR1 = 12000;
+  TIM2->CCR2 = 0;
+  TIM2->CCR3 = 12000;
+  TIM2->CCR4 = 0;
+
+
+  TIM15->CCR2 = 3000;
+  HAL_Delay(100);
+  TIM15->CCR2 = 0;
+  HAL_Delay(100);
+  TIM15->CCR2 = 3000;
+
+/*
+  TIM1->CCR1 = 0;
+  TIM1->CCR2 = 0;
+  TIM2->CCR1 = 0;
+  TIM2->CCR2 = 0;
+  TIM2->CCR3 = 0;
+  TIM2->CCR4 = 0;
+*/
+
+
+  Thrower_Send(0);
+  //wake_drivers_up();
+
+  //HAL_TIM_Base_Start_IT(&htim6);
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    /* USER CODE END WHILE */
+        wake_drivers_up();
+
+
+    /* USER CODE BEGIN 3 */
+	  if (driverReset) {
+
+	  		  driverReset = 0;
+	  		  TIM15->CCR2 = 61750;
+	  		  HAL_Delay(100);
+	  	   	  TIM15->CCR2 = 65535;
+
+	  	  }
+
+	  	  if (isCommandReceived) { // (2)
+	  	        isCommandReceived = 0;
+	  	        wake_drivers_up();
+	  	        motor1Control.speed = command.speed1;
+	  	        motor2Control.speed = command.speed2;
+	  	        motor3Control.speed = command.speed3;
+
+	  	        Thrower_Send(command.throwerSpeed);
+	    	        timer = 0;
+	    	        enable_pid = 1;
+
+	  	        feedback.speed1 = motor1Control.speed; // (4)
+	  	        feedback.speed2 = motor2Control.speed;
+	  	        feedback.speed3 = motor3Control.speed;
```

Comment:
It would be more useful to send `positionChange` with the feedback instead of the same speed values that are sent with command.